### PR TITLE
Update ScoreVariant to support exact PGS Catalog standard

### PR DESCRIPTION
**Pull request by nebfield**
## Background

We had some problems including:

* Non-additive variants caused a ValueError when instantiating a ScoreVariant
* When normalising scoring files the CLI would crash if ScoreVariants didn't look quite right
* The implementation of the ScoreVariant class wasn't great, rare combinations of fields in valid scoring files would trigger random problems

## New approach

* The pydantic model CatalogScoreVariant is an exact implementation of the PGS Catalog data standard for scoring files
* The ScoreVariant model inherits from CatalogScoreVariant to include attributes for normalisation (row_nr, accession)
* Some normalisation steps are automatically done by pydantic now (`check_effect_weight`)
* Non-additive scores are skipped by the CLI. If no scores are processed, an exception is thrown

```mermaid
---
title: PGS Catalog pydantic data models
---
classDiagram
    CatalogScoreVariant <|-- ScoreVariant
    CatalogScoreVariant : +String rsID 
    CatalogScoreVariant : +String chr_name
    CatalogScoreVariant : +int chr_position
    CatalogScoreVariant : ...
    
    note for CatalogScoreVariant "Implements PGS Catalog standard"

    class ScoreVariant{
        +String accession
        +int row_nr
        +bool is_duplicated
    }
    note for ScoreVariant "Used for normalisation"

    ScoreHeader <|-- CatalogScoreHeader

    ScoreHeader : +String pgs_id
    ScoreHeader : +Optional[String] pgs_name
    ScoreHeader : +String trait_reported
    ScoreHeader : +Optional[enum] genome_build
    ScoreHeader : +from_path()
    note for ScoreHeader "A minimal header \n for the calculator"

    note for CatalogScoreHeader "Implements PGS Catalog standard"

    class CatalogScoreHeader {
        +enum format_version
        ...
    }

    note for ScoreLog "Header metadata and \nvariant summary stats"

    class ScoreLog {
        +ScoreHeader header
        +dict variant_sources
        +bool compatible_effect_type
        ...
    }
    

```

### Gradual typing

Pydantic uses type hints to build data models, so `pgscatalog.core` now has type hints checked (if they exist) with mypy. 
  
## Test notes

Running the entire Catalog:

* `PGS002253` contains both effect weights and dosage weight columns, so I removed the restriction of exclusive weight types (if the effect weight column is present, we'll use it) 
* `PGS002263` contains a peculiar variant on row 223 

```
.       GAA     GAAA    -0.188  author_rsID=rs544624542;rs563204200;rs574206742 Unknown
```
  * This will currently raise an irrecoverable `ValidationError` because there's no positional information 
* Complex variants (e.g. APOE) may look odd, so a lot of validation is disabled for them
* Some scoring files will need their rsID field updated to have a valid prefix 
* 4 missing citations caused some ValueErrors (fixing headers)
* The new approach is slower because each variant is validated (~1 1/2 days to combine the entire Catalog), but the checks are much better 
    * I set up the CLI to work in parallel but haven't enabled this feature yet (requires changes in pgscatalog-match)
---

## Submission validation

`CatalogScoreVariant` would be a good place to implement validation logic using `field_validator` or `model_validator` so we have unified data models across submission validation and calculator normalisation. 

## Outstanding questions:

- ~~effect weights: treat them as strings that are castable to numeric, normal floats, or decimals with a defined precision? (thinking of plink limits)~~ (be consistent internally using strings that can be coerced to floats)
- [ ] Add has_complex_variants to log?

Closes https://github.com/PGScatalog/pgsc_calc/issues/370

### CODE REVIEW

File: pgscatalog.core/src/pgscatalog/core/lib/_normalise.py [hunk 1]
```diff
@@ -235,49 +170,17 @@ def remap_harmonised(variants, harmonised, target_build):
...
-                n_bad += 1
-                if not drop_missing:
-                    yield variant
-            case _:
-                yield variant
-
-    if n_bad > 1:
-        logger.warning(f"{n_bad} bad variants")
-
-
```
Comment (nebfield) [inline pgscatalog.core/src/pgscatalog/core/lib/_normalise.py:270]: This was deleted because they're mandatory fields in the pydantic model, but this will cause an immediate crash

File: pgscatalog.core/src/pgscatalog/core/lib/_normalise.py [hunk 2]
```diff
@@ -121,101 +114,43 @@ def drop_hla(variants):
...
-            case (False, True) | (None, True):
-                # none is OK because is_recessive column may be missing
-                variant.effect_type = EffectType.DOMINANT
-            case (True, False) | (True, None):
-                # none is OK because is_dominant column may be missing
-                variant.effect_type = EffectType.RECESSIVE
-            case _:
-                logger.critical(f"Bad effect type setting: {variant}")
-                raise Exception
-        yield variant
```
Comment (nebfield) [inline pgscatalog.core/src/pgscatalog/core/lib/_normalise.py:206]: This was deleted because effect_type is a computed and cached field in the pydantic model now

File: pgscatalog.core/src/pgscatalog/core/lib/_normalise.py [hunk 3]
```diff
@@ -121,101 +114,43 @@ def drop_hla(variants):
...
-    ValueError
-    """
-    for variant in variants:
-        try:
-            float(variant.effect_weight)
-        except ValueError as e:
-            logger.critical(f"{variant} has bad effect weight")
-            raise ValueError from e
-        else:
-            yield variant
```
Comment (nebfield) [inline pgscatalog.core/src/pgscatalog/core/lib/_normalise.py:145]: Handled by the pydantic model but we should think about how we handle effect weights 😅

File: pgscatalog.core/src/pgscatalog/core/__init__.py [hunk 4]
```diff
@@ -14,56 +15,20 @@
...
-    ScoreFormatError,
-    EffectTypeError,
-    CatalogError,
-    ScoreDownloadError,
-    ScoreChecksumError,
-    QueryError,
-    InvalidAccessionError,
-    SamplesheetError,
-    GenomesNotFound,
-    SamplesheetFormatError,
```
Comment (nebfield) [inline pgscatalog.core/src/pgscatalog/core/__init__.py:34]: Removed these exceptions from public API because they're only used internally by the CLI applications

File: pgscatalog.core/src/pgscatalog/core/__init__.py [hunk 5]
```diff
@@ -14,56 +15,20 @@
...
-    "ScoreFormatError",
-    "EffectTypeError",
-    "CatalogError",
-    "ScoreDownloadError",
-    "ScoreChecksumError",
-    "QueryError",
-    "InvalidAccessionError",
-    "SamplesheetError",
-    "GenomesNotFound",
-    "SamplesheetFormatError",
```
Comment (nebfield) [inline pgscatalog.core/src/pgscatalog/core/__init__.py:63]: Removed these exceptions from public API because they're only used internally by the CLI applications

**Comment by smlmbrt** [comment]
I think @ens-lgil might be a better first reviewer? And @fyvon when he gets back. Happy to look at it after.

**Comment by smlmbrt** [comment]
Re `effect_weight`: I think "strings that are castable to numeric" is the right way to go. This offloads the burden of precision to whatever downstream tool is using the information (and preserves what the author reported). Plink actually uses the full precision of the float, but rounds the output (so it's still useful information and we shouldn't arbitrarily truncate the precision).

**Comment by smlmbrt** [comment]
With regard to `pgsc_calc` the requested score will sort of disappear from the analysis after combine? Like you won't see that not of them matched, or make it into the header of the report. Is there a way to show users that it was excluded because it didn't have the right effect weights?

**Comment by fyvon** [comment]
Few questions. Not yet expert with pydantic so correct me please if I am thinking the wrong way:

1) Wouldn't this make the pgscatalog.validate package redundant? Or the validate package code would be just about catching the pydantic exceptions thrown by .core and report them in a pretty way?

2) Are we strict about scoring files formatting? For example, if I add a pattern restriction to rsID:
`rsID: Optional[str] = Field(
        default=None,
        title="dbSNP Accession ID (rsID)",
        description="The SNP’s rs ID. This column also contains HLA alleles in the standard notation (e.g. HLA-DQA1*0102) that aren’t always provided with chromosomal positions.",
        pattern=r'^rs'
    )`
  files containing ids starting with "ss" as seen recently wouldnt be compatible at all with the library, as an exception is thrown, although they would still be functionnal using chr name and position only?
  
  3) The exception mentioned in 2) is actually thrown only when I loop through the variants. Would .validate job just consist in loading all lazy-loaded fields and catching potential pydantic exceptions?

**Comment by nebfield** [comment]
> 1) Wouldn't this make the pgscatalog.validate package redundant? Or the validate package code would be just about catching the pydantic exceptions thrown by .core and report them in a pretty way?

I think yes, mostly. The original is really helpful as a reference and to provide test cases. 

A new validation CLI would import `ScoreVariant`, read an unvalidated scoring file, and try to create an object for each row. The validation CLI could be part of core or a separate package. 

>  2) Are we strict about scoring files formatting? For example, if I add a pattern restriction to rsID:
>   ```rsID: Optional[str] = Field( default=None, title="dbSNP Accession ID (rsID)", description="The SNP’s rs ID. This column also contains HLA alleles in the standard notation (e.g. HLA-DQA1*0102) that aren’t always provided with chromosomal positions.", pattern=r'^rs' )```
>        files containing ids starting with "ss" as seen recently wouldnt be compatible at all with the library, as an exception is thrown, although they would still be functionnal using chr name and position only?


Good question 👀 I left out a lot of potential validation logic. In this case [I'd do something like](https://docs.pydantic.dev/latest/concepts/validators/#field-checks):

```

@field_validator('rsid')
@classmethod
def def check_rsid_identifier(cls, v: Optional[str]) -> Optional[str]:
    if v is None:
        return None

    if v.startswith("rs") or v.startswith("ss") or v.startswith("HLA"):
        return v
    else:
       raise ValueError("rsid field must start with rs or ss or HLA")
```

And build the validation logic from collections of separate functions that are applied to individual fields or the entire model (i.e. when checking combinations of fields). 

>  3) The exception mentioned in 2) is actually thrown only when I loop through the variants. Would .validate job just consist in loading all lazy-loaded fields and catching potential pydantic exceptions?

Basically yes 🥳 The CLI could just catch, gather, and report rows and fields with problems.

**Comment by ens-lgil** [comment]
Regarding the **rsID** column, we also have the case of the "NA" value reported with the `.` character.

**Comment by nebfield** [comment]
> Regarding the **rsID** column, we also have the case of the "NA" value reported with the `.` character.

Would it be a good idea to convert this to an explicit `None` in the model? In the pydantic model I treat `NR` (e.g. in genome build) as `None` internally. 

When writing out we could convert missing values to a specific string if we need to.

**Review by fyvon — APPROVED** [review]

**Comment by nebfield** [comment]
Test notes with real data

## Failed harmonisation: PGS000012

```
....
(21988) rs1935660       A       0.012809        Unknown
(21989) rs6736922       T       0.012809        ENSEMBL rs6736922       2       219038536       A/C/G
....
```

will output a ScoreVariant with missing chr_name and chr_pos:

```
		        A		0.012809	additive	FALSE	PGS000012_hmPOS_GRCh38	21988
```

This is currently considered a valid model because the ScoreVariant has an rsid column which feels weird because harmonisation has failed. 

It's expected that this variant will never be matched against a target genome.

## Failed harmonisation: PGS000016

```
(2335) 1       1584102 A       G       2.8651e-06      Unknown
```

This is an invalid model because harmonised `None` values overwrite `chr_name` and `chr_pos` so all positional information is missing (no rsid). 

## What's the problem?

* Existing behaviour (non-pydantic) aims to keep all input variants in the output by default
* We need ValidationErrors to catch bad data and would like to make the pydantic models reusable
* Invalid pydantic models **should raise exceptions and fail to instantiate**
* So how can we write out missing data? (i.e. models that fail to instantiate)
 
Do we change the output to drop bad variants in the output by default? The log can include the variant count difference between input and output.

File: pgscatalog.core/src/pgscatalog/core/lib/models.py [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+        description="Explanation of when this variant gets included into the PGS (e.g. if it depends on the results from other variants).",
+    )
+
+    # weight information
+    # all effect weight fields are handled as strings internally to faithfully reproduce author-uploaded scores (i.e. avoid any floating point errors)
+    effect_weight: Optional[str] = Field(
+        default=None,
+        title="Variant Weight",
+        description="Value of the effect that is multiplied by the dosage of the effect allele (effect_allele) when calculating the score. Additional information on how the effect_weight was derived is in the weight_type field of the header, and score development method in the metadata downloads.",
+        coerce_numbers_to_str=True,
```
Comment (smlmbrt) [inline pgscatalog.core/src/pgscatalog/core/lib/models.py:195]: Will this cause problems with precision?

File: pgscatalog.core/src/pgscatalog/core/lib/models.py [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+    rsID: Optional[str] = Field(
+        default=None,
+        title="dbSNP Accession ID (rsID)",
+        description="The SNP’s rs ID. This column also contains HLA alleles in the standard notation (e.g. HLA-DQA1*0102) that aren’t always provided with chromosomal positions.",
+    )
+    chr_name: Optional[str] = Field(
+        default=None,
+        title="Location - Chromosome ",
+        description="Chromosome name/number associated with the variant.",
+        coerce_numbers_to_str=True,
```
Comment (smlmbrt) [inline pgscatalog.core/src/pgscatalog/core/lib/models.py:140]: Will this cause problems with precision/rounding?

**Comment by smlmbrt** [comment]
I thought the solution was to always read effect weight and position columns as str and then check that they can be coerced into a float?

**Comment by smlmbrt** [comment]
Also missing the version bump? https://github.com/PGScatalog/pygscatalog/blob/c7021b01b4774408fd1942e7928c5710ff4cfbc1/pgscatalog.core/pyproject.toml#L3

File: pgscatalog.core/src/pgscatalog/core/lib/models.py (hunk 6 above)
Comment (nebfield) [reply pgscatalog.core/src/pgscatalog/core/lib/models.py:213]: (Summary from our slack chat)

This affects instantiating the class. Without string coercion this will fail:

```python
CatalogScoreVariant(**{"effect_weight": 0.1, ...})
```

So you'd have to do:

```python
CatalogScoreVariant(**{"effect_weight": "0.1", ...})
```

Which is mildly annoying. Internally we always load weights as strings. It's up to users of this model to make sure they use strings to instantiate the class if they're worried about precision. I added a better explanation here:

https://github.com/PGScatalog/pygscatalog/blob/b379292cccbb0bff3f23590fe9c35e3e4c5e968a/pgscatalog.core/src/pgscatalog/core/lib/models.py#L102-L139
